### PR TITLE
refactor: add grantila/u2f-api (BC breaks)

### LINKED ISSUE #24: Browser support: how to check and what to do?
Two questions:

1.  What to do if a browser doesn't support U2F?
2. How do we check if a browser doesn't support U2F?

### LINKED ISSUE #25: Doesn't work with Firefox
The `u2f-api.js` is throwing:
> setting getter-only property "u2f"

Maybe an integration with https://github.com/grantila/u2f-api would be useful.

### PR DESCRIPTION
**Pull request by stephanvierkant**
Fixes #24 
Fixes #25

Warning: this is a quick and dirty proof of concept, not ready to merge.

### CODE REVIEW

File: Resources/public/js/auth.js [hunk 1]
```diff
@@ -1,107 +1,136 @@
...
+  const form = document.getElementById('u2fForm')
+  if (!form) {
+    return
+  }
+  const type = form.dataset.action
+
+  if (type === 'auth') {
+    u2fauth.authenticate()
+  } else if (type === 'reg') {
+    if (form.addEventListener) {
```
Comment (binarious) [inline Resources/public/js/auth.js]: I'd add this check in the `else if` parenthesis to decrease nesting.

File: Resources/public/js/auth.js [hunk 2]
```diff
@@ -1,107 +1,136 @@
-'use strict';
+import u2fApi from 'u2f-api'
```
Comment (binarious) [inline Resources/public/js/auth.js]: This would force people to use some sort of bundler (like webpack) to have cross-browser support.

**Comment by binarious** [comment]
Also addresses #24, doesn't it?

File: Resources/public/js/auth.js (hunk 2 above)
Comment (stephanvierkant) [reply Resources/public/js/auth.js]: That's intentionally, but opinionated ;). I assume users of this bundle use Symfony Encore, but I'm not sure if that's a good idea.

How to create a JS file that can be used with and without Webpack?

Comment (binarious) [reply Resources/public/js/auth.js]: I agree 😉. 

> I've added a package.json file, but that doesn't make any sense right now. How to solve this?

Just add a section in the readme on how to add `u2f-api`.

> How to create a JS file that can be used with and without Webpack?

The `auth` object from `auth.js` should be exported (Backwards compatible ES6 export example: https://github.com/dandv/local-iso-dt/blob/master/index.js). Then we could add a section to the readme on how to include them:

**es6-style**
```js
import u2fApi from 'u2f-api'
import auth from './auth'

auth(u2fApi);
```

**es5-style**
```js
<script src=".../u2f-api.js"></script>
<script src=".../auth.js"></script>
<script>
auth(u2fApi());
</script>
```
Just a rough idea. Maybe there are better ways.

**Comment by binarious** [comment]
@stephanvierkant Are you still on it?

**Comment by stephanvierkant** [comment]
Nope. To be honest, I totally forgot this PR.

**Comment by binarious** [comment]
Do you have the time to integrate this? (Maybe with https://github.com/darookee/u2f-two-factor-bundle/pull/35)
I'm willing to help where I can.

**Comment by stephanvierkant** [comment]
Thanks for the feedback you already gave (and for the follow-up as well). I'll try took look into this next week.

**Comment by stephanvierkant** [comment]
I'm sorry, but I'm stuck. My JS skills aren't sufficient to finish this PR to a level that it can be merged.

**Comment by binarious** [comment]
I would like to finish it if you're fine with it? Do you know if theres an option to take over a pull request?

**Comment by stephanvierkant** [comment]
Yes, please!

No idea, you can fork my fork and create a new PR? Maybe there are more elegant solutions, I'll look into this.

**Comment by binarious** [comment]
I've received your invitation, thank you. Will look into it as soon as possible.

**Comment by binarious** [comment]
Updated the PR:
- Force ES6 (Using webpack [encore] is recommended by Symfony itself).
- Added documentation
- Removed unused stuff
- Updated with latest master

@stephanvierkant @darookee Are you fine with this? It works for me in Chrome and Firefox. This introduces BC breaks (new JavaScript location and handling).

**Comment by darookee** [comment]
I'm currently swamped with work so I had no time to test this yet, but so far I'd say LGTM...

**Comment by binarious** [comment]
I'm already using this implementation in a larger app and it seems to work fine in both browsers. @stephanvierkant Did you test this as well?

**Comment by stephanvierkant** [comment]
I haven't test it yet, but I'll try it in a few days!

**Comment by binarious** [comment]
@darookee @stephanvierkant Did you find time to check and test this? I've used this implementation for more than three months in a larger app without any problems. Should be fine.

**Comment by darookee** [comment]
Another week passed and I still didn't get around to test this. @binarious are you using this exact implementation in your project? If yes I guess I just could merge and create a new release, which would be quicker, on the other hand I probably realy should test this... :-|

**Comment by binarious** [comment]
@darookee Yes, I'm still using this implementation in my application. Still no problems with different YubiKeys and different browsers.

**Comment by darookee** [comment]
I was swamped with work so I could not test this until now, looks good, just as @binarious said so I decieded to merge. Creating a new release, after checking the security related merge requests.

**Pull request merged**